### ISSUE TITLE
SoftmaxWithLoss layer for image segmentation

### ISSUE THREAD

**Issue by xujunpeng** [BVLC/caffe#5240]
Dear all,

I have encountered a problem when I setup a conv/deconv based network for image segmentation. What I did is similar to https://github.com/HyeonwooNoh/DeconvNet. The framework is like:
Input image (NxN) -> (conv layers + pool layers) -> (deconv layers + unpool layers) -> SoftmaxWithLoss 

My input are color images with 3 channels and labels are just binary (black and white) images. When I setup the SoftmaxWithLoss layer, if the softmax's bottom (Decon+ReLu) output has two or more channels, the network is working. But when the previous output only has one channel, my training is crashed.

So I am pretty confused how the softmax with loss works. Why the softmax layer cannot accept a single channel image as its bottom input?

**Comment by happyharrycn** [comment]
This is because Softmax function normalizes responses between channels and thus requires at least two channels. You should use sigmoid cross entropy loss in your case (1D output).

**Comment by xujunpeng** [comment]
Thanks happyharrycn for your reply.

So for my case,  I do have a 2D output (binarized image), I still need softmax layer right? I found https://github.com/shelhamer/fcn.berkeleyvision.org/blob/master/voc-fcn8s-atonce/net.py also used softmax layer to compute loss.

**Comment by shelhamer** [comment]
From https://github.com/BVLC/caffe/blob/master/CONTRIBUTING.md:

> _Please do not post usage, installation, or modeling questions, or other requests for help to Issues._
Use the [caffe-users list](https://groups.google.com/forum/#!forum/caffe-users) instead. This helps developers maintain a clear, uncluttered, and efficient view of the state of Caffe.
